microprofile/metrics/src/main/java/io/helidon/microprofile/metrics/MetricsCdiExtension.java

```diff
@@ -297,61 +310,124 @@ void before(@Observes BeforeBeanDiscovery discovery) {
+        if (LOGGER.isLoggable(Level.FINE)) {
+            Set<Class<?>> metricsAnnotatedClassesIgnored = new HashSet<>(metricsAnnotatedClasses);
+            metricsAnnotatedClassesIgnored.removeAll(metricsAnnotatedClassesProcessed);
+            if (!metricsAnnotatedClassesIgnored.isEmpty()) {
+                LOGGER.log(Level.FINE, () ->
+                        "Classes originally found with metrics annotations that were not processed (probably "
+                                + "because they were vetoed:" + metricsAnnotatedClassesIgnored.toString());
+            }
+        }
+        metricsAnnotatedClasses.clear();
+        metricsAnnotatedClassesProcessed.clear();
+        methodsWithSyntheticSimplyTimer.clear();
+    }
+
     /**
-     * Observes sites annotated with the metrics annotations.
+     * Records Java classes with a metrics annotation somewhere.
+     *
+     * By recording the classes here, we let CDI optimize its invocations of this observer method. Later, when we
+     * observe managed beans (which CDI invokes for all managed beans) where we also have to examine each method and
+     * constructor, we can quickly eliminate from consideration any classes we have not recorded here.
      *
-     * @param pat annotated type instance being processed
+     * @param pat ProcessAnnotatedType event
      */
-    private void registerMetrics(@Observes @WithAnnotations({Counted.class, Metered.class, Timed.class,
-                                                                    ConcurrentGauge.class, SimplyTimed.class})
-                                         ProcessAnnotatedType<?> pat) {
-        // Filter out interceptors
-        AnnotatedType<?> type = pat.getAnnotatedType();
-        Interceptor annot = type.getAnnotation(Interceptor.class);
-        if (annot != null) {
-            return;
-        }
-
-        LOGGER.log(Level.FINE, () -> "Processing annotations for " + pat.getAnnotatedType().getJavaClass().getName());
+    private void recordMetricAnnotatedClass(@Observes
+    @WithAnnotations({Counted.class, Metered.class, Timed.class, ConcurrentGauge.class,
+            SimplyTimed.class}) ProcessAnnotatedType<?> pat) {
+        checkAndRecordCandidateMetricClass(pat);
+    }
 
-        // Register metrics based on annotations
-        AnnotatedTypeConfigurator<?> configurator = pat.configureAnnotatedType();
-        Class<?> clazz = configurator.getAnnotated().getJavaClass();
+    /**
+     * Checks to make sure the annotated type is not abstract and is not an interceptor.
+     *
+     * @param pat {@code ProcessAnnotatedType} event
+     * @return true if the annotated type should be kept for potential processing later; false otherwise
+     */
+    private boolean checkCandidateMetricClass(ProcessAnnotatedType<?> pat) {
+        AnnotatedType<?> annotatedType = pat.getAnnotatedType();
+        Class<?> clazz = annotatedType.getJavaClass();
 
         // If abstract class, then handled by concrete subclasses
```

Comment:
Need to adjust this comment to match the code below it.

Reply:
Done.